sabnzbd/happyeyeballs.py

```diff
@@ -0,0 +1,111 @@
+
+import socket
+import threading
+import time
+import logging
+import queue
+from typing import NamedTuple, Tuple, Union, Optional
+
+from sabnzbd import cfg as cfg
+
+IP4_DELAY = 0.1
+
+
+# For typing and convenience!
+class AddrInfo(NamedTuple):
+    family: socket.AddressFamily
+    type: socket.SocketKind
+    proto: int
+    canonname: str
+    sockaddr: Union[Tuple[str, int], Tuple[str, int, int, int]]
+
+
+# Called by each thread
+def do_socket_connect(result_queue: queue.Queue, addrinfo: AddrInfo):
+    """Connect to the ip, and put the result into the queue"""
+    try:
+        s = socket.socket(addrinfo.family, addrinfo.type)
+        s.settimeout(3)
+
+        # Delay IPv4 connects in case we need it
+        if addrinfo.family == socket.AddressFamily.AF_INET:
+            time.sleep(IP4_DELAY)
+
+        try:
+            s.connect(addrinfo.sockaddr)
+        finally:
+            s.close()
+        result_queue.put((addrinfo, True))
+    except:
+        # We got an exception, so no successful connect on IP & port:
+        result_queue.put((addrinfo, False))
+
+
+def happyeyeballs(host: str, port: int) -> Optional[AddrInfo]:
+    """Return the fastest result of getaddrinfo()
+    based on RFC 6555 / Happy Eyeballs,
+    including IPv6 addresses if desired"""
+    try:
+        # Time how long it took us
+        start = time.time()
+
+        # Get address information, by default IPV4 or IPV6
+        family = socket.AF_UNSPEC
+        if not cfg.ipv6_servers():
+            # Only IPv4
+            family = socket.AF_INET
+        all_addrinfo = socket.getaddrinfo(host, port, family, socket.SOCK_STREAM, flags=socket.AI_CANONNAME)
+        logging.debug("Available addresses for %s (port=%d): %d", host, port, len(all_addrinfo))
+
+        # Convert to AddrInfo
+        all_addrinfo = [AddrInfo(*addrinfo) for addrinfo in all_addrinfo]
+
+        # Skip speed comparison in case we only have 1 result
+        if len(all_addrinfo) > 1:
+            # Fill queue used for threads giving back the results
+            result_queue: queue.Queue[Tuple[AddrInfo, bool]] = queue.Queue()
+            for addrinfo in all_addrinfo:
+                threading.Thread(target=do_socket_connect, args=(result_queue, addrinfo), daemon=True).start()
+
+            # start reading from the Queue for message from the threads:
+            result = None
+            for _ in range(len(all_addrinfo)):
+                connect_result = result_queue.get()
+                if connect_result[1]:
+                    result = connect_result[0]
+                    break
+        else:
+            result = all_addrinfo[0]
+
+        logging.info("Quickest IP address for %s (port=%d): %s", host, port, result.sockaddr[0])
+        logging.debug("Happy Eyeballs lookup and port connect took: %d ms", int(1000 * (time.time() - start)))
+        return result
+    except Exception as e:
+        logging.debug("Failed Happy Eyeballs lookup: %s", e)
+        return None
```

Comment:
I think we should clearly define
- what does HE return if it does not find any working connection. I would propose: None. And that we make a unity test fo that.
- and if what does the calling SAB with that. A few weeks ago I saw that SAB continues with an IP anyway. And then finds it does not work. Seems not useful to me.

Reply:
- I updated the description that indeed we only return something if we got addresses from `addrinfo` _and_ we could connect to it. So if we can't connect to any of the addresses, it will still returns none.
- Like I wrote at the top, all the other code changes in this PR make it so that your problem from a few weeks ago (#2696) doesn't happen anymore. Even if you did not set the server to Optional (that was the "solution" in #2696) the IP will _not_ be returned and the server will be ignored right away.